compiler/ballerina-lang/src/main/java/io/ballerina/projects/JarResolver.java

```diff
@@ -152,6 +157,76 @@ public Collection<JarLibrary> getJarFilePathsRequiredForExecution() {
         return jarFiles;
     }
 
+    public Collection<JarLibrary> getJarFilePathsRequiredForOptimizedExecution(HashSet<PackageId> unusedPackageIds,
+                                                                               Set<ModuleId> unusedModuleIds,
+                                                                               HashMap<PackageId, HashSet<String>> pkgWiseUsedNativeClassPaths) {
+        // 1) Add this root package related jar files
+        Set<JarLibrary> jarFiles = new HashSet<>();
+        addOptimizedCodeGeneratedLibraryPaths(rootPackageContext, PlatformLibraryScope.DEFAULT, jarFiles,
+                unusedModuleIds);
+        addUsedPlatformLibraryPaths(rootPackageContext, PlatformLibraryScope.DEFAULT, jarFiles, pkgWiseUsedNativeClassPaths);
+
+        // 2) Get all the dependencies of the root package including transitives.
+        // Filter out PackageDependencyScope.TEST_ONLY scope dependencies and lang libs
+        pkgResolution.allDependencies()
+                .stream()
+                .filter(pkgDep -> pkgDep.scope() != PackageDependencyScope.TEST_ONLY)
+                .filter(pkgDep -> !pkgDep.packageInstance().descriptor().isLangLibPackage())
+                .filter(pkgDep -> !unusedPackageIds.contains(pkgDep.packageId()))
+                .map(pkgDep -> pkgDep.packageInstance().packageContext())
+                .forEach(pkgContext -> {
+                    // Add generated thin jar of every module in the package represented by the packageContext
+                    addCodeGeneratedLibraryPaths(pkgContext, PlatformLibraryScope.DEFAULT, jarFiles);
+                    // All platform-specific libraries(specified in Ballerina.toml) having the default scope
+                    addUsedPlatformLibraryPaths(pkgContext, PlatformLibraryScope.DEFAULT, jarFiles, pkgWiseUsedNativeClassPaths);
+                });
+
```

Comment:
Between the original `getJarFilePathsRequiredForExecution` method, which is about 70 lines long, and this method, only about four lines have changed. This means there's a lot of redundant code in these two methods. You could either,
a. Introduce a new parameter( say `optimized`) in the `getJarFilePathsRequiredForExecution` method and introduce the new changes using conditions.
or
b. Move the entire step 3) to a new method.
so we can stop most of the method's logic from been repeated twice

Reply:
I think introducing a new boolean parameter will be the better approach from the two. Will work on this.

Reply:
Refactored in 21a02efd37fab58d7669956ccce6eb0aa3e93919.